File: Qb.Net/Select.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Viten.QueryBuilder.SqlOm;
using Viten.QueryBuilder.Culture;

namespace Viten.QueryBuilder
{
  /// <summary>Класс команды чтения информации</summary>
  public class Select
  {
    internal SelectQuery Query;

    internal Select(SelectQuery selectQuery)
    {
      if (selectQuery == null)
        throw new ArgumentNullException(nameof(selectQuery));
      Query = selectQuery;
    }
    internal Select(params string[] columnsName)
    {
      if (columnsName == null)
        throw new ArgumentNullException(nameof(columnsName));
      Query = new SelectQuery();
      for (int i = 0; i < columnsName.Length; i++)
        Query.Columns.Add(new SelectColumn(columnsName[i]));
    }

    internal Select(Column[] columns)
    {
      if (columns == null)
        throw new ArgumentNullException(nameof(columns));
      Query = new SelectQuery();
      for (int i = 0; i < columns.Length; i++)
        Query.Columns.Add(columns[i].Col);
    }

    bool topCalled = false;
    /// <summary>Установка параметра Top</summary>
    public Select Top(int top)
    {
      if (topCalled)
        throw new InvalidQueryException(SR.Err_RepeatTop);
      if (pageCalled)
        throw new InvalidQueryException(SR.Err_TopWithPage);
      this.Query.Top = top;
      topCalled = true;
      return this;
    }

    bool pageCalled = false;
    /// <summary>Установка постраничного просмотра</summary>
    public Select Page(int pageIndex, int pageSize)
    {
      if (pageCalled)
        throw new InvalidQueryException(SR.Err_RepeatPage);
      if (pageIndex < 0 || pageSize < 1)
        throw new InvalidQueryException(SR.Err_InvalidPage);
      if (topCalled)
        throw new InvalidQueryException(SR.Err_TopWithPage);
      this.Query.PageIndex = pageIndex;
      this.Query.PageSize = pageSize;
      pageCalled = true;
      return this;
    }

    bool distinctCalled = false;
    /// <summary>Установка параметра Distinct</summary>
    public Select Distinct(bool distinct)
    {
      if (distinctCalled)
        throw new InvalidQueryException(SR.Err_RepeatDistinct);
      this.Query.Distinct = distinct;
      distinctCalled = true;
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... FROM</summary>
    public Select From(string tableName)
    {
      if (this.Query.FromClause.BaseTable != null)
        throw new InvalidQueryException(SR.Err_RepeatFrom);
      this.Query.FromClause.BaseTable = FromTerm.Table(tableName);
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... FROM</summary>
    public Select From(string tableName, string alias)
    {
      if (this.Query.FromClause.BaseTable != null)
        throw new InvalidQueryException(SR.Err_RepeatFrom);
      this.Query.FromClause.BaseTable = FromTerm.Table(tableName, alias);
      return this;
    }

    public Select From(Union union, string alias)
    {
      if (union == null)
        throw new ArgumentNullException(nameof(union));
      if (string.IsNullOrEmpty(alias))
        throw new ArgumentException("Value should not be defined", nameof(alias));
      From from = QueryBuilder.From.Union(union, alias);
      return From(from);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="from"></param>
    /// <returns></returns>
    public Select From(From from)
    {
      if (from == null)
        throw new ArgumentNullException(nameof(from));
      if (this.Query.FromClause.BaseTable != null)
        throw new InvalidQueryException(SR.Err_RepeatFrom);
      this.Query.FromClause.BaseTable = from.Term;
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... JOIN</summary>
    public Select Join(string rightTableName, JoinCond joinCondition)
    {
      this.Query.FromClause.Join(JoinType.Inner, this.Query.FromClause.BaseTable, FromTerm.Table(rightTableName), joinCondition.Condition);
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... JOIN</summary>
    public Select Join(JoinType joinType, string leftTableName, string rightTableName, params JoinCond[] joinCondition)
    {
      From left = new From(leftTableName);
      From right = new From(rightTableName);
      return Join(joinType, left, right, joinCondition);
    }
    public Select JoinCross(string leftTableName, string rightTableName, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Cross, leftTableName, rightTableName, joinCondition);
    }
    public Select JoinFull(string leftTableName, string rightTableName, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Full, leftTableName, rightTableName, joinCondition);
    }
    public Select JoinInner(string leftTableName, string rightTableName, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Inner, leftTableName, rightTableName, joinCondition);
    }
    public Select JoinLeft(string leftTableName, string rightTableName, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Left, leftTableName, rightTableName, joinCondition);
    }
    public Select JoinRight(string leftTableName, string rightTableName, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Right, leftTableName, rightTableName, joinCondition);
    }





    /// <summary>
    /// 
    /// </summary>
    /// <param name="joinType"></param>
    /// <param name="leftTable"></param>
    /// <param name="rightTable"></param>
    /// <param name="joinCondition"></param>
    /// <returns></returns>
    public Select Join(JoinType joinType, From leftTable, From rightTable, params JoinCond[] joinCondition)
    {
      List<JoinCondition> conds = new List<JoinCondition>();
      if (joinCondition != null)
        for (int i = 0; i < joinCondition.Length; i++)
          conds.Add(joinCondition[i].Condition);
      this.Query.FromClause.Join(joinType, leftTable.Term, rightTable.Term, conds.ToArray());
      return this;
    }
    public Select JoinCross(From leftTable, From rightTable, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Cross, leftTable, rightTable, joinCondition);
    }
    public Select JoinFull(From leftTable, From rightTable, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Full, leftTable, rightTable, joinCondition);
    }
    public Select JoinInner(From leftTable, From rightTable, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Inner, leftTable, rightTable, joinCondition);
    }
    public Select JoinLeft(From leftTable, From rightTable, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Left, leftTable, rightTable, joinCondition);
    }
    public Select JoinRight(From leftTable, From rightTable, params JoinCond[] joinCondition)
    {
      return Join(JoinType.Right, leftTable, rightTable, joinCondition);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="joinType"></param>
    /// <param name="leftTable"></param>
    /// <param name="rightTable"></param>
    /// <param name="where"></param>
    /// <returns></returns>
    public Select Join(JoinType joinType, From leftTable, From rightTable, Logic where)
    {
      Join join = new Join(leftTable.Term, rightTable.Term, where.Clause, joinType);
      Query.FromClause.Joins.Add(join);
      return this;
    }

    public Select JoinInner(From leftTable, From rightTable, Logic where)
    {
      return Join(JoinType.Inner, leftTable, rightTable, where);
    }
    public Select JoinLeft(From leftTable, From rightTable, Logic where)
    {
      return Join(JoinType.Left, leftTable, rightTable, where);
    }
    public Select JoinRight(From leftTable, From rightTable, Logic where)
    {
      return Join(JoinType.Right, leftTable, rightTable, where);
    }
    public Select JoinCross(From leftTable, From rightTable, Logic where)
    {
      return Join(JoinType.Cross, leftTable, rightTable, where);
    }
    public Select JoinFull(From leftTable, From rightTable, Logic where)
    {
      return Join(JoinType.Full, leftTable, rightTable, where);
    }




    /// <summary>
    /// Все условия будут объеденены через AND
    /// </summary>
    public Select Join(JoinType joinType, From leftTable, From rightTable, params Cond[] opers)
    {
      return Join(joinType, leftTable, rightTable, Logic.And(opers));
    }
    public Select JoinCross(JoinType joinType, From leftTable, From rightTable, params Cond[] opers)
    {
      return Join(JoinType.Cross, leftTable, rightTable, Logic.And(opers));
    }
    public Select JoinFull(JoinType joinType, From leftTable, From rightTable, params Cond[] opers)
    {
      return Join(JoinType.Full, leftTable, rightTable, Logic.And(opers));
    }
    public Select JoinInner(JoinType joinType, From leftTable, From rightTable, params Cond[] opers)
    {
      return Join(JoinType.Inner, leftTable, rightTable, Logic.And(opers));
    }
    public Select JoinLeft(JoinType joinType, From leftTable, From rightTable, params Cond[] opers)
    {
      return Join(JoinType.Left, leftTable, rightTable, Logic.And(opers));
    }
    public Select JoinRight(JoinType joinType, From leftTable, From rightTable, params Cond[] opers)
    {
      return Join(JoinType.Right, leftTable, rightTable, Logic.And(opers));
    }



    /// <summary>
    /// Все условия будут объеденены через AND
    /// </summary>
    public Select Where(params Cond[] opers)
    {
      return Where(Logic.And(opers));
    }

    /// <summary>Аналог конструкции SELECT ... WHERE</summary>
    public Select Where(Logic where)
    {
      if (this.Query.WherePhrase.SubClauses.Count > 0)
        throw new InvalidQueryException(SR.Err_RepeatWhere);
      this.Query.WherePhrase.SubClauses.Add(where.Clause);
      return this;
    }

    public Select Having(params Cond[] opers)
    {
      return Having(Logic.And(opers));
    }
    /// <summary>Аналог конструкции SELECT ... HAVING</summary>
    public Select Having(Logic where)
    {
      if (this.Query.HavingPhrase.SubClauses.Count > 0)
        throw new InvalidQueryException(SR.Err_RepeatHaving);
      this.Query.HavingPhrase.SubClauses.Add(where.Clause);
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... ORDER BY</summary>
    public Select OrderBy(string field, From table, OrderByDir dir)
    {
      this.Query.OrderByTerms.Add(new OrderByTerm(field, table?.Term, dir));
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... ORDER BY</summary>
    public Select OrderBy(string field, OrderByDir dir)
    {
      this.Query.OrderByTerms.Add(new OrderByTerm(field, dir));
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... ORDER BY</summary>
    public Select OrderBy(string field)
    {
      this.Query.OrderByTerms.Add(new OrderByTerm(field, OrderByDir.Asc));
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... GROUP BY</summary>
    public Select GroupBy(string field, From table)
    {
      this.Query.GroupByTerms.Add(new GroupByTerm(field, table?.Term));
      return this;
    }

    /// <summary>Аналог конструкции SELECT ... GROUP BY</summary>
    public Select GroupBy(string field)
    {
      return GroupBy(field, null);
    }

    /// <summary>Список параметров команды</summary>
    /// <param name="parameters">Параметры</param>
    /// <returns></returns>
    public Select Params(params Param[] parameters)
    {
      if (parameters != null)
      {
        this.Query.CommandParams.AddRange(parameters);
      }
      return this;
    }

    public static explicit operator Select(SelectQuery selectQuery)
    {
      return new Select(selectQuery);
    }
  }
}

File: Qb.Js/src/Select.ts
import { SelectQuery } from "./SelectQuery";
import { Column } from "./Column";
import { From } from "./From";
import { InvalidQueryError } from "./InvalidQueryError";
import { FromTerm } from "./FromTerm";
import { Utils } from "./Utils";
import { JoinCond } from "./JoinCond";
import { JoinType, OrderByDir } from "./Enums";
import { Logic } from "./Logic";
import { Join } from "./Join";
import { OrderByTerm } from "./OrderByTerm";
import { GroupByTerm } from "./GroupByTerm";
import { Param } from "./Param";
import { Cond } from "./Cond";
import { Union } from "./Union";

/*
interface ISelect {
  Top(top: number): Select;
  Page(pageNum: number, pageSize: number): Select;
  Distinct(): Select;
  From(tableName: string): Select;
  From(tableName: string, alias: string): Select;
  From(table: From): Select;
  Join(rightTableName: string, joinCond: JoinCond): Select;
  Join(joinType: JoinType, leftTableName: From, rightTableName: From, where: Logic): Select;
  Where(where: Logic): Select;
  Having(where: Logic): Select;
  OrderBy(field: string, table: From, dir: OrderByDir): Select;
  OrderBy(field: string, dir: OrderByDir): Select;
  OrderBy(field: string): Select;
  GroupBy(field: string, table: From): Select;
  GroupBy(field: string): Select;
  Params(...parameters: Param[]): Select;
}
*/
export class Select /*implements ISelect*/ {
  private Query: SelectQuery;

  constructor(columns: Column[]) {
    this.Query = new SelectQuery();
    for (var i = 0; i < columns.length; i++)
      this.Query.Columns.push(columns[i]["Col"]);
  }

  Page(pageNum: number, pageSize: number): Select {
    this.Query.Offset = pageNum * pageSize;
    this.Query.Limit = pageSize;
    return this;
  }

  Distinct(): Select {
    this.Query.Distinct = true;
    return this;
  }

  From(tableName: string): Select;
  From(tableName: string, alias: string): Select;
  From(union: Union, alias: string): Select
  From(table: From): Select;
  From(): Select {
    if (this.Query.FromClause.BaseTable !== undefined)
      throw new InvalidQueryError("BaseTable already defined");
    var tableName: string | undefined;
    var alias: string | undefined;
    var table: From | undefined;
    var union: Union | undefined;


    if (arguments.length >= 1) {
      if (typeof arguments[0] == "string") {
        tableName = arguments[0];
      }
      else if (arguments[0] instanceof From) {
        table = arguments[0];
      }
      else if (arguments[0] instanceof Union) {
        union = arguments[0];
      }
    }
    if (arguments.length >= 2)
      alias = arguments[1];

    if (union && alias)
      table = From.Union(union, alias);

    if (table instanceof From)
      this.Query.FromClause.BaseTable = table.Term;
    else {
      if (tableName)
        this.Query.FromClause.BaseTable = FromTerm.Table(tableName, alias);
      else
        throw new InvalidQueryError("Table name not specified")
    }
    return this;
  }

  Join(rightTableName: string, joinCond: JoinCond): Select;
  Join(joinType: JoinType, leftTable: From, rightTable: From, joinCond: JoinCond): Select;
  Join(joinType: JoinType, leftTable: From, rightTable: From, where: Logic): Select;
  Join(): Select {
    var rightTableName: string | undefined;
    var joinCond: JoinCond | undefined;
    var joinType: JoinType = JoinType.Inner;
    var leftTable: From | undefined;
    var rightTable: From | undefined;
    var where: Logic | undefined;

    if (arguments.length >= 1) {
      if (typeof arguments[0] == "string") {
        rightTableName = arguments[0];
      } else {
        joinType = arguments[0];
      }
    }
    if (arguments.length >= 2) {
      if (arguments[1] instanceof JoinCond) {
        joinCond = arguments[1];
      } else {
        leftTable = arguments[1];
      }
    }
    if (arguments.length >= 3 && arguments[2] instanceof From) {
      rightTable = arguments[2];
    }
    if (arguments.length >= 4) {
      if (arguments[3] instanceof Logic) {
        where = arguments[3];
      } else if (arguments[3] instanceof JoinCond){
        joinCond = arguments[3];
      }
    }

    if (rightTableName && joinCond) {
      this.Query.FromClause.Join(joinType, this.Query.FromClause.BaseTable, From.Table(rightTableName).Term, [joinCond["Condition"]]);
    }
    else if(leftTable && rightTable && joinCond){
      this.Query.FromClause.Join(joinType, leftTable.Term, rightTable.Term, [joinCond["Condition"]]);
    } else if (leftTable && rightTable && where) {
      var join = new Join(leftTable.Term, rightTable.Term, where.Clause, joinType);
      this.Query.FromClause.Joins.push(join);
    } else {
      throw new InvalidQueryError("Invalid arguments")
    }
    return this;
  }

  Where(...conds: Cond[]): Select;
  Where(where: Logic): Select;
  Where(): Select {
    let where: Logic;
    if (arguments.length == 1 && arguments[0] instanceof Logic) {
      where = arguments[0];
    } else {
      let c: Cond[] = [];
      for (var i = 0; i < arguments.length; i++) {
        if (arguments[i] instanceof Cond) {
          c.push(arguments[i]);
        }
      }
      where = Logic.And(c);
    }
    this.Query.WherePhrase.SubClauses.push(where.Clause);
    return this;
  }

  Having(...conds: Cond[]): Select;
  Having(where: Logic): Select;
  Having(): Select {
    let where: Logic;
    if (arguments.length == 1 && arguments[0] instanceof Logic) {
      where = arguments[0];
    } else {
      let c: Cond[] = [];
      for (var i = 0; i < arguments.length; i++) {
        if (arguments[i] instanceof Cond) {
          c.push(arguments[i]);
        }
      }
      where = Logic.And(c);
    }
    this.Query.HavingPhrase.SubClauses.push(where.Clause);
    return this;
  }

  OrderBy(field: string, table: From, dir: OrderByDir): Select;
  OrderBy(field: string, dir: OrderByDir): Select;
  OrderBy(field: string): Select;
  OrderBy(): Select {
    var field: string;
    var table: From | undefined;
    var dir: OrderByDir = OrderByDir.Asc;
    if (arguments.length >= 1) {
      field = arguments[0];
    }
    if (arguments.length >= 2) {
      if (arguments[1] instanceof From) {
        table = arguments[1];
      } else {
        dir = arguments[1];
      }
    }
    if (arguments.length >= 3) {
      dir = arguments[2];
    }

    this.Query.OrderByTerms.push(new OrderByTerm(field!, dir, (table instanceof From) ? table.Term : undefined));
    return this;
  }

  GroupBy(field: string, table: From): Select;
  GroupBy(field: string): Select;
  GroupBy(): Select {
    var field: string | undefined;
    var table: From | undefined;
    if (arguments.length >= 1) {
      field = arguments[0];
    }
    if (arguments.length >= 2) {
      table = arguments[1];
    }
    this.Query.GroupByTerms.push(new GroupByTerm(field!, table ? table.Term : undefined));
    return this;
  }

  Params(...parameters: Param[]): Select {
    for (var i = 0; i < parameters.length; i++) {
      this.Query.CommandParams.push(parameters[i]);
    }
    return this;
  }
}